src/logworker.cpp

```diff
@@ -120,6 +122,14 @@ namespace g3 {
       return std::unique_ptr<LogWorker>(new LogWorker);
    }
 
+   void LogWorker::removeWrappedSink(std::shared_ptr<g3::internal::SinkWrapper> sink) {
+      auto bg_removesink_call = [this, &sink] {
```

Comment:
Naturally my findings sound weird to you :)

I made more testing and I think I can support my theory with some evidences, but to do so I am going to need your help ;)

Please follow the steps below and you will understand this issue:

1) You can take the master copy of g3log. As previously, please add sink() method and print-out of use_count in the first unit test in test_sink.cpp.
For your convinience please comment out all other unit tests, just leave this one. Once compiled and run, you should get following results (nothing new so far):

[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Sink
[ RUN      ] Sink.OneSink
handle->sink().use_count()=2
[       OK ] Sink.OneSink (505 ms)
[----------] 1 test from Sink (505 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (505 ms total)
[  PASSED  ] 1 test.

2) The next step is to add extra logging and a delay to the standart packaged_task implementation.
Please open standrad future header file (in my case it is located in /usr/include/c++/5/future, in your case it might be different).
You need to find operator () of the packaged_task template and add few lines so it looks like following:
      // Execution
      void
      operator()(_ArgTypes... __args)
      {
	__future_base::_State_base::_S_check(_M_state);
	_M_state->_M_run(std::forward<_ArgTypes>(__args)...);
// 4 lines added below
std::cout << "Before pause in packaged_task" << std::endl;
std::chrono::milliseconds wait{200};
std::this_thread::sleep_for(wait);
std::cout << "After pause in packaged_task" << std::endl;
      }

Naturally you will need to add two extra headers at the top of the file along with the rest of the headers definition:
#include iostream
#include chrono

3) Recompile your test and g3log; I had to remove content of build_travis directory to make sure I have a clean recompile.

4) After clean recompile, run test_sink and tell me what you see. If my theory is correct you should see the following print out:
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Sink
[ RUN      ] Sink.OneSink
Before pause in packaged_task
handle->sink().use_count()=3
After pause in packaged_task
Before pause in packaged_task
After pause in packaged_task
[       OK ] Sink.OneSink (505 ms)
[----------] 1 test from Sink (505 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (505 ms total)
[  PASSED  ] 1 test.

You can play a bit with pause value.

Basically it proves that if delay between setting the promise and removing context of the packaged_task is big enough (which is the case in my situation) then a valid reference to the sink is left after add / remove sink returns controls to the caller.
That is why current synchronisation based on promise / future is not sufficient in this case.
Please note that nothing is wrong with standat library, because promise / future ensure that only result from other thread is available after wait call, whereas current code relies on an assumption that not just result is available but also result producer's context is destroyed.
There are 3 options (all of them work for me) to solve it:
-> Use weak_ptr as captured value in add / remove lambdas (this is the least performant code)
-> Use empty lambdas to clear the context on the worker thread
-> Capture value by reference in add / remove lambdas (the most performant option)

Was it convincing enough for you? ;)